Add Linea Gas estimation content (#417)
Consensys/doc.zk-evm: docs/build-on-linea/gas-fees.mdx

---
description: How to estimate Linea gas costs
sidebar_position: 6
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem';

# Estimate transaction costs

Linea supports the [Ethereum EIP-1559 gas price model](https://ethereum.org/developers/docs/gas); however, being a

> layer 2 blockchain, Linea provides a more stable and cost-effective solution for transaction fees.

Reviewer: ```suggestion
layer 2 blockchain, Linea gas API provides a more accurate and cost-effective solution for transaction fees.
```

Author: This intro is trying to explain that the Linea network provides a more cost effective and stable solution to Ethereum (being a L2). I haven't yet introduces the API. However I'll be sure to highlight "accuracy" when I introduce the API.

There are minor differences in the way Linea handles gas calculations to Ethereum:

- The base fee uses a set price of 7 wei. Blocks created by Linea use up to 30 million gas
    (less than 50% of the maximum block size of 61 million gas), the fee decreases by 12.5% per block, effectively
    keeping it at a stable 7 wei.
- Transactions won't be sequenced if the `gasPrice` or `maxPriorityFeePerGas` falls below a certain threshold, which adjusts

>     over time. Instead, transactions are added to the pending queue until the gas price on the network drops sufficiently for the

Reviewer: ```suggestion
    over time and depend per transaction. Instead, transactions are added to the pending queue until the gas price on the network drops sufficiently for the
```

Author: Updated to "Transactions won't be sequenced if the `gasPrice` or `maxPriorityFeePerGas` falls below a certain threshold. This threshold is not static; it adjusts over time and varies depending on the specifics of each transaction.""

    transactions to be included.

The gas cost to submit your transaction and include it on Ethereum involves the following fee components:

- Layer 2 fee - The L2 fee (execution fee) is the cost to include your transaction on the Linea sequencer and is calculated
    using a similar formula to Ethereum. 
- Layer 1 fee - The L1 fee is the cost to publish your L2 transaction onto Ethereum and can vary based on the size
    of the transaction calldata.

You can [use the `linea_estimateGas`](../reference/api/linea-estimategas.mdx) gas API to estimate your gas costs; the result of the
call provides an estimate that includes the layer 1 and layer 2 fees.

## Calculate transaction fees

Use the [`linea_estimateGas`](../reference/api/linea-estimategas.mdx) API to estimate the gas cost for sending a

> transaction. The API will return the recommended gas limit, the base fee per gas, and the priority fee per gas.

Reviewer: I would add a note here that we still support `eth_estimateGas`, `eth_gasPrice`, `eth_feeHistory` but they will usually return a higher value, hence why we recommend using `linea_gasPrice`

:::info important

`linea_estimateGas` uses the same parameters as [`eth_estimateGas`](https://docs.infura.io/api/networks/linea/json-rpc-methods/eth_estimategas),
however we recommend using `linea_estimateGas` for more accurate results.

:::

The following example executes the `linea_estimateGas` API on the Goerli testnet.

<Tabs className="my-tabs">
  <TabItem value="curl" label="curl" default>

> ```bash

Reviewer: `bash` color coding is a bit broken in our theme. Is there a way to fix this? Maybe more a marketing thing

Author: Yes, not sure, will need to look at the configuration for that. But I think there are a few items relating to themes that may need to be adjusted in the docs.

> curl -X POST --data '{"jsonrpc": "2.0","method": "linea_estimateGas","params": [{"from": "0xFE3B557E8Fb62b89F4916B721be55cEb828dBd73","value": "0x1","gasPrice":"0x100000000","gas":"0x21000"}],"id": 53}' https://rpc.goerli.linea.build

Reviewer: We will need to update this URL to mainnet once it's ready. I don't think we need to have both.

```

</TabItem>
<TabItem value="result" label="Result">

> ```bash

Reviewer: ```suggestion
```json
```

{
  "jsonrpc": "2.0",
  "id": 53,
  "result": {
    "baseFeePerGas": "0x7",
    "gasLimit": "0xcf08",
    "priorityFeePerGas": "0x43a82a4"
  }
}
```

  </TabItem>
</Tabs>

:::note

The result of the request returns hexadecimal equivalent integers of gas prices in wei. Convert the hexadecimal value into
decimals to get the wei value. You can use any hexadecimal to decimal converter such as
[RapidTables](https://www.rapidtables.com/convert/number/hex-to-decimal.html).

:::

### Formula

The `linea_estimateGas` API calculates gas fees as follows:

- `baseFeePerGas` - Uses the Linea base fee which is static at 7 wei.
- `gasLimit` - Uses the standard `eth_estimateGas` API calculation.
- `priorityFeePerGas` - Uses the following formula:

    ```bash

>     priorityFeePerGas = L2_compressed_tx_size_in_bytes * L1_VERIFICATION_COST * L2_minimum_minable_gas_price * L1_L2_GAS_PRICE_RATIO * MINIMUM_MARGIN_ESTIMATE_GAS / (L2_tx_gas_used * L1_VERIFICATION_CAPACITY)

Reviewer: @dfox-consensys do we really want to be as transparent as this?

Reviewer 2: What about if we remove the defaults used and just describe the formula?

Author: @dfox-consensys, do you mean something like:

- `priorityFeePerGas` - Calculates the fee required to prioritize a transaction by considering factors such as the compressed transaction size, Layer 1 verification costs and capacity, gas price ratio between Layer 1 and Layer 2, the transaction's gas usage, the minimum minable gas price on Layer 2, and a minimum margin (for error) for gas price estimation."

Reviewer 2: Yes, LGTM

---
description: How to estimate Linea gas costs
sidebar_position: 6
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem';

# Estimate transaction costs

Linea supports the [Ethereum EIP-1559 gas price model](https://ethereum.org/developers/docs/gas); however, being a
layer 2 blockchain, Linea provides a more stable and cost-effective solution for transaction fees.

There are minor differences in the way Linea handles gas calculations to Ethereum:

- The base fee uses a set price of 7 wei. Blocks created by Linea use up to 30 million gas

>     (less than 50% of the maximum block size of 61 million gas), the fee decreases by 12.5% per block, effectively

Reviewer 3: ```suggestion
    (less than 50% of the maximum Linea block size of 61 million gas), and the fee decreases by 12.5% per block, effectively
```

    keeping it at a stable 7 wei.
- Transactions won't be sequenced if the `gasPrice` or `maxPriorityFeePerGas` falls below a certain threshold. This threshold
    is not static; it adjusts over time and varies depending on the specifics of each transaction. Instead, transactions
    are added to the pending queue until the gas price on the network drops sufficiently for the transactions to be included.

The gas cost to submit your transaction and include it on Ethereum involves the following fee components:

- Layer 2 fee - The L2 fee (execution fee) is the cost to include your transaction on the Linea sequencer and is calculated
    using a similar formula to Ethereum. 
- Layer 1 fee - The L1 fee is the cost to publish your L2 transaction onto Ethereum and can vary based on the size
    of the transaction calldata.

You can [use the `linea_estimateGas`](../reference/api/linea-estimategas.mdx) gas API to estimate your gas costs; the result of the
call provides an estimate that includes the layer 1 and layer 2 fees.

## Calculate transaction fees

> Use the [`linea_estimateGas`](../reference/api/linea-estimategas.mdx) API to estimate the gas cost for sending a
> transaction. The API will return the recommended gas limit, the base fee per gas, and the priority fee per gas.

Reviewer 3: ```suggestion
Use the [`linea_estimateGas`](../reference/api/linea-estimategas.mdx) API method to estimate the gas cost for sending a
transaction. The method returns the recommended gas limit, the base fee per gas, and the priority fee per gas.
```

:::info important

Linea also supports [`eth_estimateGas`](https://docs.infura.io/api/networks/linea/json-rpc-methods/eth_estimategas),
[`eth_gasPrice`](https://docs.infura.io/api/networks/linea/json-rpc-methods/eth_gasprice), and
[`eth_feeHistory`](https://docs.infura.io/api/networks/linea/json-rpc-methods/eth_feehistory), but they usually return
a higher (and less accurate) estimate than `linea_estimateGas`.

:::

> The following example executes the `linea_estimateGas` API on the Linea Goerli testnet.

Reviewer 3: ```suggestion
The following example executes `linea_estimateGas` on the Linea Goerli testnet.
```

<Tabs className="my-tabs">
  <TabItem value="curl" label="curl" default>

```bash
curl -X POST --data '{"jsonrpc": "2.0","method": "linea_estimateGas","params": [{"from": "0xFE3B557E8Fb62b89F4916B721be55cEb828dBd73","value": "0x1","gasPrice":"0x100000000","gas":"0x21000"}],"id": 53}' https://rpc.goerli.linea.build
```

</TabItem>
<TabItem value="result" label="Result">

```json
{
  "jsonrpc": "2.0",
  "id": 53,
  "result": {
    "baseFeePerGas": "0x7",
    "gasLimit": "0xcf08",
    "priorityFeePerGas": "0x43a82a4"
  }
}
```

  </TabItem>
</Tabs>

The `linea_estimateGas` API calculates gas fees as follows:

> - `baseFeePerGas` - Uses the Linea base fee which is static at 7 wei.

Reviewer 3: ```suggestion
- `baseFeePerGas` - Uses the Linea base fee which is set at 7 wei.
```

- `gasLimit` - Uses the standard `eth_estimateGas` API calculation.
- `priorityFeePerGas` - Calculates the fee required to prioritize a transaction by considering factors such as the
    compressed transaction size, layer 1 verification costs and capacity, gas price ratio between layer 1 and

>     layer 2, the transaction's gas usage, the minimum minable gas price on Layer 2, and a minimum

Reviewer 3: ```suggestion
    layer 2, the transaction's gas usage, the minimum gas price on layer 2, and a minimum
```
